SmartDeviceLink/private/SDLProtocolReceivedMessageProcessor.m

```diff
@@ -0,0 +1,317 @@
+                default:
+                    prevState=state;
+                    state = ERROR_STATE;
+                    break;
+            }
+            break;
+            
+        case CONTROL_FRAME_INFO_STATE:
+            // 8 bits for frame information
+            controlFrameInfo = currentByte;
+            
+            // Set the next state
+            state = SESSION_ID_STATE;
+            
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Check for errors.
+            // For these two frame types, the frame info should be 0x00
+            if ((frameType == SDLFrameTypeFirst) || (frameType == SDLFrameTypeSingle)){
+                if (controlFrameInfo != 0x00) {
+                    prevState=state;
+                    state = ERROR_STATE;
+                }
+            }
+            break;
+            
+        case SESSION_ID_STATE:
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Set the next state
+            state = DATA_SIZE_1_STATE;
+            break;
+        
+        // 32 bits for data size
+        case DATA_SIZE_1_STATE:
+            dataLength = 0;
+            dataLength += (currentByte & 0xFF ) << 24;
+            state = DATA_SIZE_2_STATE;
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            break;
+            
+        case DATA_SIZE_2_STATE:
+            dataLength += (currentByte & 0xFF ) << 16;
+            state = DATA_SIZE_3_STATE;
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            break;
+            
+        case DATA_SIZE_3_STATE:
+            dataLength += (currentByte & 0xFF ) << 8;
+            state = DATA_SIZE_4_STATE;
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            break;
+            
+        case DATA_SIZE_4_STATE:
+            dataLength += (currentByte & 0xFF ) << 0;
+            
+            // Set the next state
+            state = MESSAGE_1_STATE;
+            
+            // Add the byte to the headerBuffer
+            [self.headerBuffer appendBytes:&currentByte length:sizeof(currentByte)];
+            
+            // Set the counter for the data pump.
+            dataBytesRemaining = dataLength;
+            
+            // Version 1 does not have a message ID so we skip to the data pump, or the end.
+            if( version == 1){
+                if (dataLength == 0) {
+                    [self ResetState];
+                } else {
+                    // skip ahead to the data pump state
+                    state = DATA_PUMP_STATE;
```

Comment:
Should this break?

Reply:
No.  We set the state, then do our error checking.  If we find an error that state will change.

Reply:
I was wrong.  On review of it, yes, because the latest version calls the reset function, this should break.